tests: logging: log_api: Skip message dropping test for SMP

nordic-krch opened the pull request:
When SMP is used other core may process logging and we won't
get expected number of dropped messages. Skipping test when
SMP is enabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: logging: log_api: Execute message dropping test on 1 cpu

Ensure that test is executed on 1 CPU. When target has multiple
CPUs other core may process logging and we won't get expected number
of dropped messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

dcpleung commented on tests/subsys/logging/log_api/src/test.inc:
```diff
@@ -591,6 +591,13 @@ static void log_n_messages(uint32_t n_msg, uint32_t exp_dropped)
  */
 static void test_log_msg_dropped_notification(void)
 {
+	if (IS_ENABLED(CONFIG_SMP)) {
```
Testing for `CONFIG_MP_NUM_CPUS > 1` might be better, since you can have only 1 CPU under SMP (this is usually used to test SMP code path).

andyross replied:
Alternatively use a _1cpu_ ztest test case here, which will lock down all but one CPU while the test runs (via a terrible hack, but it does work).  Definitely would be sad to lose test coverage on SMP systems.

andyross replied:
Gah, formating: `_1cpu_`

nordic-krch replied:
@andyross it's using this new ztest API and i'm not sure what's the best way to run 1cpu test so i used @dcpleung suggestion. @yperess how to setup a single test case for 1 cpu?

yperess replied:
Yes, the best way is to create a new suite. The reason is that generally speaking you'll be adding more things to test additional code paths or edge cases to that suite. The thinking is to encourage small tests. In a way this is similar to adding `{}` to `if` statements even when there's just one statement. It's setting up for future changes.

One thing you could do if you wanted to avoid different test suites is to create a test rule that checks the test name. Something along the lines of if the test name ends with `_1cpu` then run the 1CPU start/stop functions.

andyross reviewed (changes requested):
I hate to say it, but... I think I'm going to -1 this patch.  It's just not right to lose code coverage on some platforms simply because the test system made something that used to be easy harder.

Reading the new API, it looks to me like the before/after function hooks are now part of a test "suite" and not an individual case.  But 1cpu wrapper hooks are still there and declared in the same header.  Can you maybe try putting the offending case into a "single core" suite instead?

nordic-krch committed:
tests: logging: log_api: Reworking suite setup

Reworking test suite setup.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
testsuite: ztest: Fix concatenation in macro

Use UTIL_CAT which resolves input arguments instead of direct
preprocessor concatenation.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@andyross i agree. I reworked suite setup so that this test is executed on 1 cpu only. Ideally it would be good to have something like `ZTEST_1CPI()` or test specific setup and teardown like we have in deprecated style (@yperess any ideas?). As for now i had to create dedicated suite only for that one test case.
